src/librustc/mir/interpret/mod.rs

```diff
@@ -238,6 +242,127 @@ pub fn specialized_decode_alloc_id<
     }
 }
 
+// Used to avoid infinite recursion when decoding recursive allocations.
+pub type DecodingSessionId = NonZeroU32;
+
+// FIXME: This could be represented as an AtomicU64, letting us get rid of the
+//        Mutex in AllocIdDecoder
+#[derive(Copy, Clone)]
+enum State {
+    Empty,
+    InProgress(DecodingSessionId, AllocId),
+    Done(AllocId),
```

Comment:
If you made this store the Result instead of just the Ok value, you can get rid of the panicking and poisoning

Reply:
I'll look into it.